lib/vtls/mesalink.c

```diff
@@ -173,22 +173,24 @@ mesalink_connect_step1(struct connectdata *conn, int sockindex)
       failf(data,
             "WARNING: failed to configure server name indication (SNI) "
             "TLS extension\n");
+      return CURLE_SSL_CONNECT_ERROR;
     }
-    else {
-      /* Check if the hostname is 127.0.0.1 or [::1];
-       * otherwise reject because MesaLink always wants a valid DNS Name
-       * specified in RFC 5280 Section 7.2 */
-      if(strncmp(hostname, "127.0.0.1", 9) == 0
+  }
+  else {
+    /* Check if the hostname is 127.0.0.1 or [::1];
+     * otherwise reject because MesaLink always wants a valid DNS Name
+     * specified in RFC 5280 Section 7.2 */
+    if(strncmp(hostname, "127.0.0.1", 9) == 0
 #ifdef ENABLE_IPV6
-         || strncmp(hostname, "[::1]", 5) == 0
+       || strncmp(hostname, "[::1]", 5) == 0
 #endif
-      ) {
-        SSL_set_tlsext_host_name(BACKEND->handle, "localhost");
-      }
-      else {
-        failf(data,
-              "ERROR: MesaLink does not accept an IP address as a hostname\n");
-      }
+    ) {
+      SSL_set_tlsext_host_name(BACKEND->handle, "localhost");
```

Comment:
Why this nasty kludge? Surely there's no support in any standard for this behavior? No name == no SNI!

Reply:
I don't like this nasty thing, either! Fine, I'll implement no-SNI support in MesaLink...

Reply:
No name == no SNI. At least this is true for our dependency `webpki`, which we use to verify certificate. Please see the API signature below. It always requires a DNS name as specified in RFC 5280 Section 7.2.

https://github.com/briansmith/webpki/blob/68fcf3dd3b7f90fc51ebde1397da7b6f5d1e4034/src/webpki.rs#L193-L197

Reply:
SNI is a TLS *extension*. If you connect to a host without having a known name that we can't pass one for SNI. It is not an uncommon situation. All other TLS libraries can handle this and just don't set any SNI then.

If your library can't handle a non-SNI handshake then I suggest you simply return an error if there's no name to use - which then makes it impossible to connect to "https://192.168.0.1" etc - even if the user disables cert checking. Don't invent a host name, that's plain wrong.

Reply:
I agree it is plain wrong to invent a host name. So the current backend only allows 127.0.0.1 and [::1] and rejects other IP addresses.

https://github.com/curl/curl/blob/8c65cbd575e3994fc44daaa24960c7dc04e92384/lib/vtls/mesalink.c#L190-L194

Reply:
I agree with you. I don't like inventing hostnames either. Anyhow `webpki` always requires a valid DNS name to verify certificates. To make `webpki` happy, I would have to make a hostname somewhere; either in the MesaLink library or in the vtls backend. The `localhost` kludge is a compromise to make both `webpki` and curl's test cases happy...

Reply:
Since you can't negotiate TLS without SNI (which is remarkably non-openssl compliant), you must fail the connection. Than yes, you can't run most of the curl test suite. I can't accept that you put that logic into curl for release code.

We probably then need to add some required feature-bit in the TLS-tests that require non-SNI connects to work so that curl+mesalink skips those. A mesalink build could then use `#ifdef CURLDEBUG` and make up the host name accordingly for testing purposes only so that debug builds can run through more tests.